parsers/c/c_lexer.mll

```diff
@@ -152,10 +145,56 @@ let cn_keywords: (string * Tokens.token) list = [
     "implies"       , CN_IMPLIES;
   ]
 
-let cn_lexicon: (string, token) Hashtbl.t =
-  let cn_lexicon = Hashtbl.create 0 in
-  let add (key, builder) = Hashtbl.add cn_lexicon key builder in
-  List.iter add cn_keywords; cn_lexicon
+(* CN keywords that are functional, but have limitations that make them
+unsuitable for non-experts *)
+let cn_keywords_experimental: (string * Tokens.token) list = [
+    "cn_list"       , CN_LIST; 
+    "cn_tuple"      , CN_TUPLE;
+    "cn_map"        , CN_MAP;
+    "cn_set"        , CN_SET;
+    "cn_have"       , CN_HAVE;
+    "cn_function"   , CN_FUNCTION;
+    "cn_print"      , CN_PRINT;
+  ] 
+
+(* CN keywords that are non-functional, but reserved for possible future use *)
+let cn_keywords_unimplemented: (string * Tokens.token) list = [
+    "pack"          , CN_PACK;
+    "unpack"        , CN_UNPACK;
+  ]
+
+let cn_lex_builder kw_list : (string, token) Hashtbl.t  = 
+  let cn_lex = Hashtbl.create 0 in
+  let add (key, builder) = Hashtbl.add cn_lex key builder in
+  List.iter add kw_list; cn_lex
+
+(* 
+Attempt to lex a CN keyword. These may be: 
+  * 'production' - well-supported and suitable for general use
+  * 'experimental' - functional in some cases but not recommended for general use 
+  * 'unimplemented' - non-functional, but the keyword is reserved  
+*)
+let cn_lex_keyword id start_pos end_pos = 
+  (* Try to lex CN production keywords *)
+  try Hashtbl.find (cn_lex_builder cn_keywords) id
+  with Not_found ->
+    (* Try to lex CN experimental keywords - warn on success *)
+    try 
+      let kw = Hashtbl.find (cn_lex_builder cn_keywords_experimental) id in 
+      prerr_endline       
+        (Pp_errors.make_message
+         Cerb_location.(region (start_pos, end_pos) NoCursor)
+         Errors.(CPARSER (Errors.Cparser_experimental_keyword id))
+         Warning);
+      kw 
+    with Not_found ->
+      try 
+        (* Try to lex CN unimplemented keywords - throw an error on success *)
+        let _ = Hashtbl.find (cn_lex_builder cn_keywords_unimplemented) id in 
+        raise (Error (Errors.Cparser_unimplemented_keyword id))
+      with Not_found ->
+        UNAME id
```

Comment:
I wonder whether the three-level lookup will be slow. If it is, we may want to have a single keywords list of type `(string * (Tokens.token * (normal+experimental+unimplemented) ) list`

Reply:
I haven't benchmarked this, but this seems like a small source of inefficiency so I'm inclined to let it go. Would you prefer if I refactored it?

Reply:
Actually, I think your suggestion is better - I've implemented it.